Adding cloud storage capabilities

brettp opened the pull request:
DO NOT MERGE. This is an initial commit for early feedback from @Elgg/contributors and MITRE. There is a lot left to address:
- [ ] Access race condition when deleting ElggEntity with metadata.
- [x] ~~Elgg disk caching solutions~~ Unrelated. We still need a cache dir.
- [ ] Numerous core / bundled plugins access files directly
- [ ] Need unified thumbnail generation on ElggFile. See #6152.
- [x] Pull cloud adapters out into plugin.
- [ ] Add better tests
- [ ] Fixup and rebase before merge

brettp committed:
feature(system): Added ElggCloudFilestore with an AwsS3 adapater and relevant config.

Refs #3766

ewinslow commented:
Couple initial thoughts:

* Do you have any usage examples? What is the API for plugin devs?

* I'd like to avoid adding more classes without a namespace (I'm looking at `ElggCloudFilestore`).

* Lets also try to avoid introducing more global state (i.e. `elgg_get_config` and other global function friends). I believe we have services for most of these things now, so you can inject those in the constructor

* In general, I find `new` in anything but the DI system to be a smell unless you're constructing an in-memory-only data structure.

* `Elgg\Filesystem\*` was initially designed to be divorced from most of Elgg and only provide a thin wrapper around Gaufrette so we could switch it out for something else if we ever wanted to ditch Gaufrette for some reason (e.g. for Flysystem). In particular this means any of our code that returns a plain Gaufrette\Filesystem is "leaking" implementation details. We can either try to hide Gaufrette behind Filesystem or we can just give up and let Gaufrette completely define our filesystem abstraction, but I wouldn't want a mixed approach.

* Is there a plan on what testing will look like?

brettp commented:
Sorry for the noise on this. I had my branch in a munged state. :-X

> Do you have any usage examples? What is the API for plugin devs?

This adds an option in admin to set where you want user data stored, and then registering that as the default filestore. It's transparent to plugin devs if they're using ElggFile. We'll need to add some new methods for it to cover streams, though. I'm not convinced a direct API should be available to plugin devs.

> Elgg\Filesystem\* was initially designed to be divorced from most of Elgg and only provide a thin wrapper around Gaufrette so we could switch it out for something else if we ever wanted to ditch Gaufrette for some reason (e.g. for Flysystem). In particular this means any of our code that returns a plain Gaufrette\Filesystem is "leaking" implementation details. We can either try to hide Gaufrette behind Filesystem or we can just give up and let Gaufrette completely define our filesystem abstraction, but I wouldn't want a mixed approach.

For strictly local disk operations or anything Gaufrette supports? Right now Elgg\Filesystem is pretty tightly bound to Gaufrette. Elgg\Filesystem\GaufretteDirectory is the only way to access directories. If the idea is inject Gaufrette adapters and use Elgg\Filesystem\* for everything, it'll need work. 

ElggFile and the filestore concept don't lend themselves too well to our newer APIs and practices. How much are we looking to change for 1.x? Anything as long as BC is maintained?

brettp commented:
Ah, testing: I'm looking at that next. As I'm getting into this, I'm seeing there's more design decisions that need rethought than I expected, so wanted to have this conversation about making those changes before starting to add S3 tests.

ewinslow commented:
> For strictly local disk operations or anything Gaufrette supports? 

It should support anything supported by Gaufrette.

> Right now Elgg\Filesystem is pretty tightly bound to Gaufrette. Elgg\Filesystem\GaufretteDirectory is the only way to access directories. If the idea is inject Gaufrette adapters and use Elgg\Filesystem* for everything, it'll need work.

If you look at the Elgg\Filesystem\Directory interface it makes no references to Gaufrette. As long as you only type-hint on that, you should be able to nearly ignore Gaufrette outside of Elgg\Filesystem namespace. `GaufretteDirectory` is an implementation that adds a few static factory methods for generating appropriate adapters without getting your hands too dirty with Gaufrette's internals and such.

This is a helpful discussion a lot of these things I have thought through in my head but not communicated well outside. It might be beneficial to turn the results of this discussion into docs and shove them onto the classes at least.

ewinslow commented:
> ElggFile and the filestore concept don't lend themselves too well to our newer APIs and practices. How much are we looking to change for 1.x? Anything as long as BC is maintained?

Yes, I'm up for anything (i.e. rewrite) as long as we keep BC. Then we can decide later what to do in 2.0 (e.g. just drop the cruft?

brettp committed:
feature(system): Using DI for data storage.

Refs #3766

brettp commented:
Still a WIP, but I've updated the approach. Take a look and let me know any feedback.

* Added Elgg\DataStorageService and Elgg\Filesystem\AdapterService to manage the default and available data stores and how they interact with the physical data.
* Rewriting ElggFile and the filestore system. 
* The API wrapping is [public] ElggFile -> [public?] Elgg\DataStorageService([private] Elgg\Filesystem\Adapter) -> [private] Gaufrette

Example DataStorage API:
```php
$storage = _elgg_services()->dataStorage;
$adapters = _elgg_services()->dataStorageAdapters;

// requires the aw3_s3 configured through admin.
$storage->setAdapter($a->get('aws_s3'));
// disk is always set
// $storage->setAdapter($a->get('disk'));

$storage->file('filename')->setContent('test data');
var_dump($s->file('filename')->getContent());

var_dump($storage->directory('some/directory')->listKeys());

$storage->directory('another/directory')->file('another_file')->setContent('more data');
```

The ElggFile API looks the same.

Things left to consider:
* Storing the data storage adapter info on the Elgg File object
* Exposing proxy settings

brettp commented:
@Elgg/contributors - Any comments on this? I'll be off this project in a few weeks so would like to get as much done as possible before then.

brettp committed:
chore(cloud): removed old cloud filestore.

ewinslow commented on engine/classes/Elgg/DataStorageService.php:
```diff
@@ -0,0 +1,40 @@
+		$this->logger = $logger;
+	}
+	
+	public function setAdapter(Adapter $adapter) {
```
I'd recommend just using constructor params instead of setters. We probably don't want the adapter changing mid-request anyways, and this class has no reasonable behavior without the adapter configured.

Actually, I'm not sure I understand what this class is for. The API is no different than the adapters, really.

brettp replied:
> I'd recommend just using constructor params instead of setter
Services don't support constructor arguments.

> Actually, I'm not sure I understand what this class is for. The API is no different than the adapters, really.
Hmm...I don't remember why I separated these out now, either. I'll double check my notes on Monday at work, but I agree the two APIs can probably be combined.

ewinslow replied:
What do you mean that services don't support arbitrary constructor arguments?

brettp replied:
You can't pass constructor arguments to a service from _elgg_service() because it's already instantiated. Maybe I don't understand what you're suggesting.

ewinslow replied:
I'm suggesting we instantiate this class with all the needed dependencies.

ewinslow commented on engine/classes/Elgg/Filesystem/Adapter/AwsS3.php:
```diff
@@ -0,0 +1,48 @@
+
+class AwsS3 extends Filesystem implements Adapter {
+	private $config;
+	public function __construct(array $config) {
```
This constructor does a lot of work. Why not just ask for a ready-to-go GaufretteAwsS3 adapter?

brettp replied:
The work has to be done somewhere, and I thought it was cleaner to do it here than in ElggFile. Doing it here also keeps Gaufrette further away from the public API.

ewinslow replied:
I think the typical place for this now is in the dependency injector or some factory function. I'd much prefer to keep our constructors really dumb (field assignments only). If we do that I think we may even be able to drop another class in this case, too.

ewinslow replied:
Some more background on where I'm coming from:

http://misko.hevery.com/code-reviewers-guide/flaw-constructor-does-real-work/

ewinslow commented on engine/classes/Elgg/Filesystem/Adapter/AwsS3.php:
```diff
@@ -0,0 +1,48 @@
+use Gaufrette\Adapter\AwsS3 as GaufretteAwsS3;
+use Aws\S3\S3Client;
+
+class AwsS3 extends Filesystem implements Adapter {
```
I've grown to treat inheritance with suspicion. Is that really what we mean to do here?

brettp replied:
Yes, I meant to inherit from Filesystem to avoid either re-implementing or wrapping the entire Gaufrette class.

ewinslow commented on engine/classes/Elgg/Filesystem/AdapterService.php:
```diff
@@ -0,0 +1,96 @@
+		return isset($this->adapters[$name]);
+	}
+	
+	public function remove($name) {
```
delete? remove() in the other collections APIs will take the value to be removed, not the key.

ewinslow replied:
$coll->add($item);
$coll->remove($item);

vs.

$map->set($key, $item);
$map->delete($key); // or unset?

brettp replied:
Which ones? The only one I see that takes a value is ElggPriorityList. All the other ones take ->remove($key) or ->delete($key). It's not a huge deal since there needs to be documentation, but I'd rather follow current convention than introduce more inconsistent behavior.

ewinslow replied:
This is the one I had in mind: https://github.com/Elgg/Elgg/pull/7941/files#diff-6ee5e38e8c654ac5e5f240855531ca32R18

brettp replied:
I just saw those classes this morning. It's a shame they're not used in core.

I'll be able to switch over to ArrayCollection assuming plugins are fully init'd by the time the service is called.

ewinslow replied:
Need to actually merge them in first :). I will make an extra effort to get
them in this week.

On Mon, Apr 6, 2015 at 12:15 PM Brett Profitt <notifications@github.com>
wrote:

> In engine/classes/Elgg/Filesystem/AdapterService.php
> <https://github.com/Elgg/Elgg/pull/8092#discussion_r27821409>:
>
> > +	 * @return Adapter
> > +	 * @throws \InvalidArgumentException
> > +	 */
> > +	public function get($name) {
> > +		if (!$this->has($name)) {
> > +			throw new \InvalidArgumentException("Filesystem Adapter \"$name\" does not exist.");
> > +		}
> > +		
> > +		return $this->adapters[$name];
> > +	}
> > +	
> > +	public function has($name) {
> > +		return isset($this->adapters[$name]);
> > +	}
> > +	
> > +	public function remove($name) {
>
> I just saw those classes this morning. It's a shame they're not used in
> core.
>
> I'll be able to switch over to ArrayCollection assuming plugins are fully
> init'd by the time the service is called.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/8092/files#r27821409>.
>

ewinslow commented on engine/classes/Elgg/Filesystem/AdapterService.php:
```diff
@@ -0,0 +1,96 @@
+		return array_keys($this->adapters);
+	}
+	
+	public function setDefault($name) {
```
If default can only be set once, then we should do it at construction somehow rather than throwing exceptions. Or perhaps the default setting shouldn't be part of this class and we just need to make this a dumb map instead.

brettp replied:
The default adapter isn't available until after `init, system`, so we can't set it in the services as a constructor argument. I'm leaning toward removing the default stuff from this class and requiring explicit calls for adapter types. The default filestore concept looks like it was to keep from crashing if the filestore info changed, so I'm not convinced it's the right solution.

ewinslow commented on engine/classes/ElggFile.php:
```diff
@@ -233,28 +298,51 @@ public function write($data) {
+	 *
+	 * @return mixed The file contents.
+	 */
+	public function grabFileOld() {
```
All these *Old methods are making me think we should use some good old polymorphism instead:

ElggFileOld vs ElggFileNew classes, all the methods stay the same -- grabFile, read, etc.

Better names would be good, but you get the idea.

brettp replied:
Ahh...meant to remove *Old before the commit.

ewinslow replied:
Ah ok. Deleting code is even better!

ewinslow commented on engine/classes/Elgg/Filesystem/AdapterService.php:
```diff
@@ -0,0 +1,96 @@
+		return true;
+	}
+	
+	public function listAdapters() {
```
If we just call this keys() then we are one step closer to just using a standard Map structure instead of something custom.

ewinslow commented on engine/classes/Elgg/Filesystem/AdapterService.php:
```diff
@@ -0,0 +1,96 @@
+	public static function buildFromParams(array $params) {
+		// build and return a new adapter from a set of param
+		// expects an array with classname and adapter-specific params
+		if (!isset($params['classname'])) {
```
Should classname be passed as a second function argument?

ewinslow replied:
It is pulled out of $params three times here, and something tells me that it isn't normal for a class to take its own name as a parameter, so I doubt it needs to be passed to the new statement below.

brettp replied:
Yeah, I don't think I like this either. The idea here is to save the adapter info as MD on the ElggFile object so it can be re-instantiated correctly. I understand the point of this (multiple data stores), but don't think it has a real use case.

ewinslow replied:
I'd certainly like to avoid class names in the database. We do that with subtypes already and I think it causes some unnecessary refactoring pain.

brettp replied:
What do you think about removing support for each ElggFile having its own datastore and only using the store configured in admin? It would remove the "default" datastore logic and saving classnames / adapter names in the DB. The down side is switching datastores would require some sort of migration.

ewinslow commented on engine/classes/ElggCloudFilestore.php:
```diff
@@ -0,0 +1,189 @@
+	public function __construct($adapter = null) {
+		if ($adapter) {
+			$this->adapter = $adapter;
+			$this->fs = $adapter->getFilesystem();
```
This class is going to blow up if it is used without fs being set. What was the motivation for nullable adapter here?

ewinslow replied:
Also, why not inject the filesystem directly? Law of Demeter seems to apply, yea?

brettp replied:
Accidental commit. Disregard this file.

ewinslow commented on engine/classes/ElggFile.php:
```diff
@@ -361,6 +456,32 @@ public function exists() {
+	 * 
+	 * @return \Elgg\Filesystem\Adapter\Adapter
+	 */
+	public function getDataStorage() {
```
This getter feels a bit spooky. Has side effects on global state?

Why do users of ElggFile need to use it to get a data storage adapter?

brettp replied:
Same deal as above -- To let the adapter save info in MD to re-instantiate.

ewinslow commented on engine/tests/phpunit/Elgg/DataStorageServiceTest.php:
```diff
@@ -0,0 +1,45 @@
+		$this->adapter = $this->getMock('Elgg\Filesystem\Adapter\Adapter', [], [], '', false);
+	}
+	
+	public function testCanSetAdapter() {
```
This test makes no assertions.

ewinslow commented on engine/tests/phpunit/Elgg/Filesystem/AdapterServiceTest.php:
```diff
@@ -0,0 +1,188 @@
+	/**
+	 * @dataProvider multipleAdapters
+	 */
+	function testGet($s, $adapters) {
```
Can you be explicit about what behavior you're testing for? Even if it feels like you're just returning the PHPdoc for the get method?

ewinslow commented on views/default/forms/admin/site/advanced/system.php:
```diff
@@ -33,5 +33,81 @@
+		<p class="elgg-text-help"><?php echo elgg_echo('admin:settings:user_data:aws_s3:info'); ?></p>
+		<?php
+			$info = elgg_get_config('user_data_store_info');
+			$s3 = elgg_extract('aws_s3', $info, []);
```
Do we want to bake Aws3 support into core or just "allow" it in an "Amazon services" type plugin?

I feel uneasy being tied to particular third parties, no matter how popular. It is often too hard to keep up with their pace of change.

brettp replied:
Wouldn't mind putting the specific implementations into a plugin. This goes with #8111.

ewinslow commented on engine/lib/filestore.php:
```diff
@@ -418,10 +418,14 @@ function _elgg_clear_entity_files($entity) {
  */
 function get_default_filestore() {
+	_elgg_services()->deprecation
+		->sendNotice("Use Elgg\Filesystem\AdapterService::setDefault() instead of get_default_filestore()", 1.11);
```
getDefault?

ewinslow commented on engine/lib/filestore.php:
```diff
@@ -429,19 +433,24 @@ function get_default_filestore() {
 /**
  * Set the default filestore for the system.
+ * 
+ * @depreciated 1.11 Use Elgg\Filesystem\AdapterService::setDefault() or set $CONFIG->user_data_store = 'adapter_name'
```
I don't think we want to encourage setting `$config` settings directly anymore.

ewinslow commented on views/default/forms/admin/site/advanced/system.php:
```diff
@@ -33,5 +33,81 @@
+				'value' => elgg_extract('key', $s3)
+			]);
+			
+			$secret_label = elgg_echo('admin:settings:user_data:aws_s3:secret');
```
Can we suffix the "label" language keys with "label" so that there is room for other things in the future like "placeholder", "description", etc.?

ewinslow commented:
OK, left a bunch of comments, but I think the most needful thing here is really just documentation. In particular: goals/non-goals of this effort. Should plugins be able to add their own remote storage options or is the idea that core would just support a few of the popular options (aws/disk at first)? This code gives me the impression it's the latter but I think the general philosophy of Elgg is more the former. I don't recall what we talked about before anymore... :(

Also: Justifications for the existence of the various APIs. "This is needed in order to..." type php doc comments would be great. Example use-cases are great.

juho-jaakkola commented:
> Should plugins be able to add their own remote storage options or is the idea that core would just support a few of the popular options (aws/disk at first)?

I'd prefer doing it through plugins.

brettp commented:
Thanks for the feedback. I agree that specific implementations should be put into plugins, though that doesn't answer the question of what, if any, do we want to officially support as bundled plugins.

ewinslow commented on engine/classes/ElggDiskFilestore.php:
```diff
@@ -430,4 +430,8 @@ protected function makeFileMatrix($guid) {
 	}
+	
+	public function moveUploadedFile($from, $to) {
+		return move_uploaded_file($from, $to);
```
How does this behave on a non local time system?

brettp replied:
It'll need to be re-implemented for a non-local system using is_uploaded_file().

ewinslow commented on engine/lib/filestore.php:
```diff
@@ -463,11 +472,35 @@ function elgg_get_file_simple_type($mime_type) {
+	$config = _elgg_services()->config;
+	
+	// register core fs adapters
+	$adapters = _elgg_services()->dataStorageAdapters;
```
I imagine all this logic can go into the dataSorageAdapters di factory function. Then we don't need elgg_filestore_init at all.

ewinslow replied:
On there's some plugins hooks that have to stick around, but it greatly simplifies things in any case

ewinslow commented:
A couple more general comments:

* Type hinting wherever possible would be great.
* I'd love to move towards a more immutable by default approach with our services. So they are ready to go at construction and can't be modified mid request. This makes a bunch of potential bugs impossible and allows us to skip throwing and catching variations of "object not initialized" exceptions.

brettp committed:
feature(system): Changed API for data storage.
Fixed problems with directories and saving files.
Combined DataStorageService with AdapterService.

Refs #8092

brettp committed:
feature(cloud): moved cloud storage options into plugin.
Added Azure and Google Cloud.

brettp commented:
I changed a number of approaches in the latest commit and would appreciate feedback. Some highlights:

* Pulled the guts of the cloud adapters out into a plugin. The plugin is currently in core, but I'm thinking it needs to be a separate repo so composer deps resolve correctly.
* Removed concept of a default filestore and ElggFile entities saving their filestore information to DB.
* Simplified the StorageAdapterService by using straight Gaufrette adapters.
* Registering the selected storage adapter as a stream wrapper for `elgg://dataStorage` so we can pass around streams.
* Added support for MS Azure and Google Cloud Storage

ewinslow commented on engine/lib/filestore.php:
```diff
@@ -418,10 +418,14 @@ function _elgg_clear_entity_files($entity) {
 /**
  * Return the default filestore.
+ * 
+ * @depreciated 1.11 Use _elgg_services()->dataStorage
```
depreciated != deprecated :)

ewinslow commented on mod/cloud_data_storage/classes/Elgg/Filesystem/Adapter/MsAzure/IServiceElggFilter.php:
```diff
@@ -0,0 +1,53 @@
+ * A filter for Azure SDK REST requests that optionally uses a proxy
+ * or disables SSL validation
+ */
+class IserviceElggFilter implements IServiceFilter {
```
S should be capitalized in this name,

ewinslow commented on mod/cloud_data_storage/classes/Elgg/Filesystem/Adapter/MsAzure/IServiceElggFilter.php:
```diff
@@ -0,0 +1,53 @@
+
+    public static function errorHandler($errno, $errorMessage, $errorFile, $errorLine)
+    {
+        return ($errno == E_WARNING ? true : false);
```
No need for the ternary.

ewinslow commented on mod/cloud_data_storage/composer.json:
```diff
@@ -0,0 +1,21 @@
+{
+    "name": "elgg/elgg-cloud-storage",
```
elgg/cloud-data-storage? None of the core elgg repos repeat elgg on the right hand side of the name.

ewinslow commented on mod/cloud_data_storage/composer.json:
```diff
@@ -0,0 +1,21 @@
+    "name": "elgg/elgg-cloud-storage",
+    "version": "1.11.0-dev",
+    "description": "Adds cloud data storage to Elgg.",
+    "license": "GPL-2.0",
```
Can we make this MIT?

ewinslow commented on mod/cloud_data_storage/composer.json:
```diff
@@ -0,0 +1,21 @@
+{
+    "name": "elgg/elgg-cloud-storage",
+    "version": "1.11.0-dev",
```
Drop the version

ewinslow commented on mod/cloud_data_storage/views/default/admin/site/advanced/storage_aws_s3.php:
```diff
@@ -0,0 +1,52 @@
+?>
+<p class="elgg-text-help"><?php echo elgg_echo('admin:settings:user_data_store:aws_s3:info'); ?></p>
+<div>
+	<label>
```
Thank you for making this form more accessible than the Elgg standard :)

ewinslow commented:
Agree on moving plugin to separate repo.

ewinslow commented:
Thanks for adding more adapters. Screenshots of what the end-user will see may help... Forgive me for my laziness in setting up a running instance of this.
